doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,258 @@
+ * @param {(context: ModuleExportsContext) => {ModuleExportsResult}} nextExports
+ * @returns {ModuleExportsResult}
+ */
+function exports(url, context, nextExports) {  // Mocking require-in-the-middle
+  let { exports: originalExports } = nextExports(url, context);
+  const stats = getStats(url);
+  if (stats.name && modules.includes(stats.name)) {
+    const newExports = userCallback(originalExports, stats.name, stats.basedir);
+    return { exports: newExports }
+  }
+  return {
+    exports: originalExports
+  };
+}
+```
+
+This can only be meaningfully implemented for modules loaded by `require()`. ESM's design is completely different from CJS in that resolution of the graph and evaluation of the graph are separated, so a similar timing in ESM would be "after module evaluation". However, Node.js only gets to control the timing of the evaluation of the root module. The inner module evaluation is currently completely internal to V8. It may be possible to upstream a post-evaluation hook to V8, but calling from C++ to JS would also incur a non-trivial performance cost, especially if it needs to be done for every single module. Also, since the ESM namespace is specified to be immutable, what users can do after ESM evaluation is very limited - they cannot replace anything in the namespace or switch it to a different module. That's why the `link` hook was devised below to better work with the design of ESM (before linking completes it's possible to swap the module resolved by `import` to a different module).
+
+### Alternative design: `requires` that encompasses `resolve` and `load`
+
+An alternative design would be to span it across the `resolve` and `load` steps - take the specifier as argument, return exports in the result, and rename it to something like `requires()` (to avoid clashing with `require()`). The `nextRequires` hook would for example, invoke the default implementation of `requires` which in turn encompasses `resolve` and `load`. If `nextRequires` is not invoked then the default `resolve` and `load` will be skipped.
+
+```js
+function requires(specifier, context, nextRequires) {  // Mocking require-in-the-middle
+  let { exports: originalExports, url } = nextRequires(specifier, context);
+  const stats = getStats(url);
+  if (stats.name && modules.includes(stats.name)) {
+    const newExports = userCallback(originalExports, stats.name, stats.basedir);
+    return { exports: newExports }
+  }
+  return {
+    exports: originalExports
+  };
+}
+```
+
+It's not yet investigated whether it is possible to make it work with the CJS loader cache at all, but it looks closer to what developers generally try to monkey-patch the CJS loader for.
+
+## `link` (import-only): invoked before linking
+
+This is invoked after `load` but prior to Node.js passing the final result to V8 for linking, so it can be composed with `resolve` and `load` if necessary. This needs to work with experimental vm modules for passing module instances around.
```

Comment:
Since this is a completely new hook and not encumbered by the existing hooks API, how about "preLink" for the name? To me "link" means that this hook performs some sort of linking.

Reply:
IMO `beforeLink` looks a bit more intuitive

Reply:
Actually the V8 API for this is `instantiate`. Although I am not sure if it's better than `translate`, but at least both sound better than `link`, maybe except for people who read ESM specs..

Reply:
Changed this to `instantiate`, because `link` actually spans compile + instantiation (dependency resolving) and I think for the purpose of this hook, it should only just be for instantiation (that's the earliest point to get the exports parsed by V8, compile is actually too early because there can be `export * from '...'` which won't be available until after dependency resolution, which is why V8 only allows querying of the namespace after instantiation).